File: web-app/src/encoding.ts
// TypeScript port of chronael/encoding.py. MUST match it bit-for-bit, or the model
// receives garbage. Verified against the Python encoding (same planes, same move
// indices, same chosen moves) before shipping.
//
// Board is oriented so the side to move is always "White": when Black is to move we
// mirror vertically and swap colours (python-chess board.mirror()). Square indices
// follow python-chess: index = rank*8 + file, with rank 0 = rank 1, file 0 = file a.
// square_mirror(sq) = sq ^ 56 flips the rank.

import type { Chess } from "chess.js";

export const NUM_PLANES = 17;
export const POLICY_SIZE = 64 * 64 * 5; // 20480

const PIECE_TYPE: Record<string, number> = { p: 1, n: 2, b: 3, r: 4, q: 5, k: 6 };
const PROMO_CODE: Record<string, number> = { n: 1, b: 2, r: 3, q: 4 };

function squareIndex(sq: string): number {
  const file = sq.charCodeAt(0) - 97; // 'a' -> 0
  const rank = sq.charCodeAt(1) - 49; // '1' -> 0
  return rank * 8 + file;
}

/** Encode a FEN into the (17,8,8) planes, flattened as plane*64 + rank*8 + file. */
export function encodePlanes(fen: string): Float32Array {
  const [placement, turn, castling, ep] = fen.split(" ");
  const flipped = turn === "b";
  const planes = new Float32Array(NUM_PLANES * 64);

  // Piece placement: FEN lists rank 8 first (row 0) down to rank 1 (row 7).
  const rows = placement.split("/");
  for (let row = 0; row < 8; row++) {
    let file = 0;
    for (const ch of rows[row]) {
      if (ch >= "1" && ch <= "8") {
        file += ch.charCodeAt(0) - 48;
        continue;
      }
      const isWhite = ch === ch.toUpperCase();
      const t = PIECE_TYPE[ch.toLowerCase()];
      let rank = 7 - row; // python-chess rank (0 = rank 1)
      let colorWhite = isWhite;
      if (flipped) {
        rank = 7 - rank;
        colorWhite = !isWhite;
      }
      const plane = t - 1 + (colorWhite ? 0 : 6);
      planes[plane * 64 + rank * 8 + file] = 1;
      file++;
    }
  }

  // Castling rights, read as the side-to-move (White) after orientation.
  const wk = castling.includes("K");
  const wq = castling.includes("Q");
  const bk = castling.includes("k");
  const bq = castling.includes("q");
  const rights = flipped ? [bk, bq, wk, wq] : [wk, wq, bk, bq];
  rights.forEach((has, i) => {
    if (has) {
      const base = (12 + i) * 64;
      for (let s = 0; s < 64; s++) planes[base + s] = 1;
    }
  });

  // En-passant target square.
  if (ep && ep !== "-") {
    const file = ep.charCodeAt(0) - 97;
    let rank = ep.charCodeAt(1) - 49;
    if (flipped) rank = 7 - rank;
    planes[16 * 64 + rank * 8 + file] = 1;
  }

  return planes;
}

export interface LegalMove {
  from: string;
  to: string;
  promotion?: string;
}

/** Legal moves and their oriented-frame policy indices (to gather model logits). */
export function legalMoveIndices(chess: Chess): { moves: LegalMove[]; indices: number[] } {
  const flipped = chess.turn() === "b";
  const moves = chess.moves({ verbose: true }) as unknown as LegalMove[];
  const indices = moves.map((m) => {
    let from = squareIndex(m.from);
    let to = squareIndex(m.to);
    if (flipped) {
      from ^= 56; // square_mirror
      to ^= 56;
    }
    const code = m.promotion ? PROMO_CODE[m.promotion] ?? 0 : 0;
    return (from * 64 + to) * 5 + code;
  });
  return { moves, indices };
}

File: chronael/encoding.py
"""Board and move encoding for the Chronael policy network.

Design goals
------------
* **Side-to-move orientation.** The board is always presented to the network with
  the side to move at the bottom (as if it were White). When it is Black's turn we
  mirror the board (vertical flip + colour swap). This lets the network learn "my
  pieces vs. their pieces" instead of "white vs. black", which roughly halves the
  amount it has to learn and makes data augmentation unnecessary.

* **Injective move <-> index mapping.** Every legal chess move maps to a unique
  index in ``[0, POLICY_SIZE)`` via ``(from*64 + to)*5 + promo_code``. This is a
  superset of all legal moves, so decoding is never needed: at inference we simply
  gather the logits of the *legal* moves and pick among them. The model therefore
  can never output an illegal move.
"""

from __future__ import annotations

import chess
import numpy as np

# 6 piece types x 2 colours = 12 piece planes, + 4 castling, + 1 en-passant.
NUM_PLANES = 17
BOARD_SIZE = 8

# Move index space: 64 from-squares * 64 to-squares * 5 promotion codes.
_PROMO_TO_CODE = {None: 0, chess.KNIGHT: 1, chess.BISHOP: 2, chess.ROOK: 3, chess.QUEEN: 4}
_CODE_TO_PROMO = {v: k for k, v in _PROMO_TO_CODE.items()}
POLICY_SIZE = 64 * 64 * 5  # 20480


def orient(board: chess.Board) -> tuple[chess.Board, bool]:
    """Return ``(oriented_board, flipped)`` where the side to move is always White.

    ``flipped`` is True when the original side to move was Black, in which case the
    board has been mirrored vertically with colours swapped.
    """
    if board.turn == chess.WHITE:
        return board, False
    return board.mirror(), True


def orient_move(move: chess.Move, flipped: bool) -> chess.Move:
    """Map a move into / out of the oriented frame.

    ``chess.square_mirror`` is its own inverse, so the same call orients a move for
    encoding and un-orients a model-chosen move back to the real board.
    """
    if not flipped:
        return move
    return chess.Move(
        chess.square_mirror(move.from_square),
        chess.square_mirror(move.to_square),
        promotion=move.promotion,
    )


def move_to_index(move: chess.Move) -> int:
    """Map an (already oriented) move to its unique policy index."""
    return (move.from_square * 64 + move.to_square) * 5 + _PROMO_TO_CODE[move.promotion]


def index_to_move(index: int) -> chess.Move:
    """Inverse of :func:`move_to_index` (used for tests / debugging only)."""
    promo_code = index % 5
    squares = index // 5
    to_sq = squares % 64
    from_sq = squares // 64
    return chess.Move(from_sq, to_sq, promotion=_CODE_TO_PROMO[promo_code])


def board_to_planes(oriented_board: chess.Board) -> np.ndarray:
    """Encode an *oriented* board (side to move == White) into ``(17, 8, 8)`` floats."""
    planes = np.zeros((NUM_PLANES, BOARD_SIZE, BOARD_SIZE), dtype=np.float32)

    for square, piece in oriented_board.piece_map().items():
        rank = chess.square_rank(square)
        file = chess.square_file(square)
        plane = (piece.piece_type - 1) + (0 if piece.color == chess.WHITE else 6)
        planes[plane, rank, file] = 1.0

    # Castling rights (side to move == White after orientation).
    if oriented_board.has_kingside_castling_rights(chess.WHITE):
        planes[12, :, :] = 1.0
    if oriented_board.has_queenside_castling_rights(chess.WHITE):
        planes[13, :, :] = 1.0
    if oriented_board.has_kingside_castling_rights(chess.BLACK):
        planes[14, :, :] = 1.0
    if oriented_board.has_queenside_castling_rights(chess.BLACK):
        planes[15, :, :] = 1.0

    # En-passant target square.
    if oriented_board.ep_square is not None:
        rank = chess.square_rank(oriented_board.ep_square)
        file = chess.square_file(oriented_board.ep_square)
        planes[16, rank, file] = 1.0

    return planes


def encode_position(board: chess.Board) -> tuple[np.ndarray, bool]:
    """Encode any board for the network. Returns ``(planes, flipped)``."""
    oriented, flipped = orient(board)
    return board_to_planes(oriented), flipped


def legal_move_indices(board: chess.Board) -> tuple[list[chess.Move], list[int], bool]:
    """For a position, return its legal moves, their policy indices, and ``flipped``.

    The returned moves are in the *original* board frame (ready to push), while the
    indices are in the oriented frame (matching the network's output).
    """
    oriented, flipped = orient(board)
    moves: list[chess.Move] = []
    indices: list[int] = []
    for move in board.legal_moves:
        moves.append(move)
        indices.append(move_to_index(orient_move(move, flipped)))
    return moves, indices, flipped
